### ISSUE TITLE
Analyze the performance of the new way of obtaining agents groups

### ISSUE THREAD

**Issue by GGP1** [wazuh/wazuh#19426]
## Description

In https://github.com/wazuh/wazuh/issues/18675 we introduced some changes to the way be obtain agents' groups and we noticed that there is a considerable performance degradation ([1](https://github.com/wazuh/wazuh/issues/18675#issuecomment-1739003716), [2](https://github.com/wazuh/wazuh/issues/16285#issuecomment-1584748806)) when trying to obtain a large number of agents.

This issue aims to review the performance of the new implementation and look for potential optimizations to improve it.

## Checks

The following elements have been updated or reviewed (should also be checked if no modification is required):
- [ ] Tests (unit tests, API integration tests).
- [ ] Changelog.
- [ ] Documentation.
- [ ] Integration test mapping (using `api/test/integration/mapping/_test_mapping.py`).

**Comment by GGP1** [comment]
## Update

I executed several queries on two branches with and without the changes to compare the performances. The environment contained 50,000 agents.

<details><summary>group column in the agent table</summary>

#### Getting agent id and group

```sql
sqlite> SELECT id, group FROM agent;
Run Time: real 0.077 user 0.015834 sys 0.059156
Run Time: real 0.080 user 0.016771 sys 0.061670
Run Time: real 0.074 user 0.019248 sys 0.053991
```

#### Getting all fields

```sql
sqlite> SELECT * FROM agent;
Run Time: real 0.286 user 0.097132 sys 0.100948
Run Time: real 0.272 user 0.114902 sys 0.082904
Run Time: real 0.273 user 0.103036 sys 0.094959
```

</details>

<details><summary>group column in the belongs table</summary>

#### All agents belonging to the 'default' group.

```sql
sqlite> SELECT agent.id, GROUP_CONCAT(belongs.name_group) FROM agent LEFT JOIN belongs ON agent.id=belongs.id_agent GROUP BY agent.id;
Run Time: real 0.124 user 0.047194 sys 0.074557
Run Time: real 0.117 user 0.031795 sys 0.083412
Run Time: real 0.125 user 0.047883 sys 0.076117
``` 

#### All agents belonging to the 'default' and 'group1' groups.

```sql
sqlite> SELECT agent.id, GROUP_CONCAT(belongs.name_group) FROM agent LEFT JOIN belongs ON agent.id=belongs.id_agent GROUP BY agent.id;
Run Time: real 0.143 user 0.063709 sys 0.075553
Run Time: real 0.129 user 0.061748 sys 0.066900
Run Time: real 0.131 user 0.056536 sys 0.072810
``` 

#### All agents belonging to the 'default', 'group1' and 'group2'.

```sql
sqlite> SELECT agent.id, GROUP_CONCAT(belongs.name_group) FROM agent LEFT JOIN belongs ON agent.id=belongs.id_agent GROUP BY agent.id;
Run Time: real 0.163 user 0.058070 sys 0.091434
Run Time: real 0.144 user 0.062464 sys 0.077832
Run Time: real 0.147 user 0.068118 sys 0.076262
``` 

#### All agents belonging to the three groups and getting all agent fields

```sql
sqlite> SELECT agent.*, GROUP_CONCAT(belongs.name_group) FROM agent LEFT JOIN belongs ON agent.id=belongs.id_agent GROUP BY agent.id;
Run Time: real 0.389 user 0.181182 sys 0.140682
Run Time: real 0.384 user 0.217519 sys 0.107053
Run Time: real 0.379 user 0.197894 sys 0.122233
```

</details>

|  Two fields  | 4.8.0 (No 1NF) | #17127 (1NF) |
|----|---------------------|---------------------|
|0   | 0.077s             | 0.163s             |
|1   | 0.080s             | 0.144s             |
|2   | 0.074s             | 0.147s             |
|**Mean**| **0.077s** | **0.15133333333333s (+65.1095%)**|

|  All fields   | 4.8.0 (No 1NF) | #17127 (1NF) |
|----|---------------------|---------------------|
|0   | 0.286s             | 0.389s             |
|1   | 0.272s             | 0.384s             |
|2   | 0.273s             | 0.379s             |
|**Mean**| **0.277s** | **0.384s (+32.3752%)**|

As we can see from the results, retrieving just the `id` and `group` takes 65% more time in the new structure, and it takes longer if we add more groups, since each one of the group represents a new row.

Selecting all fields is closer in `real` time `0.286 (old)` vs `0.389 (new)`, the difference is only ~1/4 of the time, but it takes approximately twice as much `user` (CPU) time `0.097132 (old)` vs `0.181182 (new)` in the new implementation.

> `user` indicates the cumulative time spent by all the CPUs during the computation.

To sum up, there is a significant degradation in the performance of the queries and a higher resource consumption with the new architecture but it's not near what has been [reported](https://github.com/wazuh/wazuh/issues/18675#issuecomment-1739003716) in the epic issue. I will keep investigating to find what is causing the problem.

**Comment by GGP1** [comment]
## Update

I've found that the performance difference between both implementations when calling the `GET /agents` endpoint isn't too concerning (+12.66%). 

However, in the new one, if we specify the `limit` parameter and set a value of `50000` it then takes around 9  to 10 seconds more to execute. Still not the values reported in the epic issue but significantly higher than the other request.

I measured the time taken by the SQL query with the `LIMIT` value set and it's almost identical to the one without it. This means that there's an issue with the Wazuh's agent logic when the parameter `limit` is specified, but I couldn't find any evidence of where the problem is yet.

> All the tests were perfomed in an environment with 50k agents, each one belonging to three groups.

### WazuhDBQueryAgents execution time

```python
start_time = time.time()
with WazuhDBQueryAgents(offset=offset, limit=limit, sort=sort, search=search, select=select,
                                query=q, **rbac_filters, distinct=distinct) as db_query:
    data = db_query.run()

logger.info(str(time.time() - start_time))
```

| | 4.8.0 | 4.8.0 (#17127) |
| -- | -- | -- |
| 0 | 0.023s | 0.097s |
| 1 | 0.022s | 0.095s |
| 2 | 0.022s | 0.089s |
| 3 | 0.025s | 0.091s |
| 4 | 0.021s | 0.089s |
| 5 | 0.021s | 0.092s |
| 6 | 0.021s | 0.089s |
| 7 | 0.030s | 0.088s |
| 8 | 0.022s | 0.090s |
| 9 | 0.021s | 0.097s |
| Mean | **0.023s** | **0.092s (+300%)** |

### GET /agents request execution time

| | 4.8.0 | 4.8.0 (#17127) |
| -- | -- | -- |
| 0 | 542ms | 672ms |
| 1 | 571ms | 637ms |
| 2 | 569ms | 615ms |
| 3 | 560ms | 628ms |
| 4 | 554ms | 598ms |
| Mean | **559.2ms** | **630ms (+12.66%)** |

### WazuhDBQueryAgents execution time with a limit = 50000

| | Time |
| -- | -- |
| 0 | 8.70s | 
| 1 | 8.65s | 
| 2 | 8.56s | 
| 3 | 8.52s | 
| 4 | 9.01s |
| Mean | **8.69s** |

### GET /agents?limit=50000 request execution time

| | Time | 
| -- | -- |
| 0 | 9.98s |
| 1 | 9.88s |
| 2 | 9.74s |
| 3 | 9.85s |
| 4 | 10.16s |
| Mean | **9.922s** |

> Difference with the old structure is **+1674.95%**

### SQL performance of the query used to get all agents and a limit of 50000

```sql
sqlite> SELECT agent.*, GROUP_CONCAT(belongs.name_group) FROM agent LEFT JOIN belongs ON agent.id=belongs.id_agent GROUP BY agent.id LIMIT 50000 OFFSET 0;
Run Time: real 0.384 user 0.187112 sys 0.159832
Run Time: real 0.394 user 0.190678 sys 0.158148
Run Time: real 0.473 user 0.203537 sys 0.151729
```

**Comment by GGP1** [comment]
## Update

I've measured the execution times taken by the Wazuh framework to process a `GET /agents?limit=50000` request with 50,000 agents each one belonging to three groups and I've noticed that the bottleneck is in the execution of the query by the backend. 


https://github.com/wazuh/wazuh/blob/dd354194d55f4ad116631cadf5399808a8034613/framework/wazuh/core/utils.py#L1674

https://github.com/wazuh/wazuh/blob/dd354194d55f4ad116631cadf5399808a8034613/framework/wazuh/core/utils.py#L1354-L1357

### Framework benchmarks

| | `_execute_data_query` | `general_run` | `WazuhDBQueryAgents` | Request roundtrip |
| -- | -- | -- | -- | -- |
| 0 | 5.57s | 5.61s | 6.41s | 7.60s |
| 1 | 5.48s | 5.52s | 6.37s | 7.46s |
| 2 | 5.60s | 5.65s | 6.43s | 7.66s |
| 3 | 5.44s | 5.48s | 6.37s | 7.58s |
| 4 | 5.44s | 5.49s | 6.27s | 7.48s |
| **Mean** | **5.50s** | **5.55s** | **6.37s** | **7.55s** |

### SQLite benchmarks

I ran the exact same query in the master's node local database and it performs well, it doesn't seem to be an issue with how the query is constructed.

```sql
sqlite> SELECT os_uname as 'os.uname',group_config_status as 'group_config_status',manager_host as 'manager',os_major as 'os.major',os_platform as 'os.platform',last_keepalive as 'lastKeepAlive',internal_key as 'internal_key',os_version as 'os.version',version as 'version',os_name as 'os.name',node_name as 'node_name',merged_sum as 'mergedSum',os_minor as 'os.minor',date_add as 'dateAdd',config_sum as 'configSum',coalesce(ip,register_ip) as 'ip',os_build as 'os.build',register_ip as 'registerIP',os_codename as 'os.codename',connection_status as 'status',os_arch as 'os.arch',disconnection_time as 'disconnection_time',name as 'name',GROUP_CONCAT(belongs.name_group) as 'group',id as 'id' FROM agent LEFT JOIN belongs ON id=belongs.id_agent GROUP BY id ORDER BY id ASC LIMIT 50000 OFFSET 0;
```

| | real | user | sys |
| -- | -- | -- | -- |
| 0 | 0.346s | 0.195s | 0.093s |
| 1 | 0.324s | 0.137s | 0.124s |
| 2 | 0.318s | 0.179s | 0.093s |
| 3 | 0.317s | 0.151s | 0.115s |
| 4 | 0.316s | 0.145s | 0.109s |
| **Mean** | **0.324s** | **0.161s** | **0.106s** |

Just to discard a common problem, the `belongs` table columns are indexed, so there shouldn't be an issue with that neither.

https://github.com/wazuh/wazuh/blob/08381cb95895fd0cdfe191b83b881246a74385e4/src/wazuh_db/schema_global.sql#L66-L75

### Conclusion

The results indicate that there must be an issue in the socket that the framework uses to communicate with the backend or in the backend itself when retrieving the results from the database.

**Comment by fdalmaup** [comment]
# Review 

The analysis was correctly carried out, comparing the queries directly made to the DB and the ones via the Wazuh Framework. The comparison results concluded that the normalization itself did not generate the degradation of performance in the queries to the database but rather the logic used by `WazuhDBBackend` is not processing the obtained results efficiently.

The team should discuss the possible approaches to take regarding this matter since the `WazuhDBBackend` class is used for many other database queries.
